doc/manual/src/contributing/experimental-features.md

```diff
@@ -0,0 +1,59 @@
+This section describes the notion of *experimental features*, and how it fits into the big picture of the development of Nix.
+
+# What are experimental features?
+
+[Since Nix 2.4](../release-notes/rl-2.4.md), some features of Nix can be flagged as experimental.
+Experimental features can be changed or removed at any time, and users must [explicitly enable them](../nix-conf.md#conf-experimental-features).
+This allows accessing new functionality that is not yet stable without unwittingly relying on them.
+
+# When should a new feature be marked experimental?
+
+A change in the Nix codebase should be guarded by an experimental feature flag if it is considered likely to be reverted or adapted in a backwards-incompatible manner after gathering more experience with it in practice.
+
+Examples:
+
+- Changes to the Nix language, such as new built-ins, syntactic or semantic changes, etc.
+- Changes to the command-line interface
```

Comment:
```suggestion
A change in the Nix codebase should be guarded by an experimental feature flag if it needs the Nix community to gather experience with it before we can decide what we want it to be.

For most changes, experimenting with the pull request branch before it is merged should be good enough --- if the Nix maintainers are not confident in the design or implementation, then the PR must not be merged. But for some larger, highly visible, or important changes, this will not work:

  - The Nix maintainers need input from community at large need to help with the decision. The Nix community at large needs an easy and safe way to experiment and gain experience with the idea, and trying out pull request miscellaneous branches is neither easy or safe enough.

  - A long period of experimentation and gaining experience is necessary. Keeping a PR open for that length of time would result in and undue burden of fixing conflicts. With just a PR, we also loose out on the opportunity for the change (especially if the maintainers are leaning towards accepting it) influencing other development and thus gaining experience with the maintained costs and benefits of the features.
  
Conversely, just merging the feature and hoping it was designed correctly the first time is also impractical:
 - If we don't want the feature after all, removing it is a breaking change and prohibited without a major version bump. 
 - Even if we do mostly like the feature it may still some breaking changes to improve its design, again prohibited without a major version bump.

What we thus need is a middle-ground between merging a feature which instantly becomes stable, and requesting further changes to the PR without merging it. That is where experimental features come in:

 - The feature is merged so it is easy to try it out and gain more experience
 - The feature is not stable so breaking changes to the feature including removal are completely permissible
 - The feature is *guarded* so regular uses will not use it by mistake, and thus are shielded from any instability.

Examples:

- Changes to the Nix language, such as new built-ins, syntactic or semantic changes, etc.
- Changes to the command-line interface
```

Reply:
I think we've already captured that in [one sentence:](https://github.com/NixOS/nix/pull/5930/files#diff-c60361f52f0e428ab378856cac200ca405b6f43f71a1c78621c67208ca4e3391R53)
> An experimental feature enables developers to iterate on and deliver a new idea without committing to it or requiring a costly long-running fork.